### PR TITLE
acts: ensure Python_EXECUTABLE uses ^python when +python

### PR DESCRIPTION
**Pull request by wdconinc**
By default, `find_package(Python)` searches from highest version to lowest version, identifying the highest version that satisfies the requirements. This means that `/usr/bin/python3.11` will be found before `$(spack location -i python)/bin/python3.10`, even when other packages have been built with the `python` in spack.

This ensures that the `python` dependency is explicitly the `python` version that is used by `acts`.

### CODE REVIEW

**Comment by HadrienG2** [comment]
This sounds like something that has broader applicability than Acts and should somehow live in the Python package rather than the Acts package. In an Acts context specifically, doing it for the acts package only could mean that the Python version used by Acts on one hand and ROOT + dd4hep on the other hand could go out of sync, which is undesirable.

However, I'm not sure if Spack already has the infrastructure to add CMake flags to downstream packages like that. Pinging a few core spack devs for thoughts: @adamjstewart @alalazo .

**Comment by alalazo** [comment]
@HadrienG2 I think we can make this a method of `CMakeBuilder`, so we can encode in that function the behavior from different CMake versions. In a recipe a client would just:
```python
args.extend(self.find_python(self.spec["python"]))
```

**Review by alalazo — APPROVED** [review]

**Comment by alalazo** [comment]
@HadrienG2 Waiting for your approval before merging.

**Comment by HadrienG2** [comment]
@alalazo Would it be possible to somehow modify the python spackage to ensure that any CMake package with a depends_on("python") automatically gets the right CMake flag added ? This would be the best behavior here, as it would ensure that any package that depends on python actually gets the expected python version from spack, and not whatever FindPackage(Python) feels like finding today.

**Comment by alalazo** [comment]
@HadrienG2 I think it's feasible to do that, but the code belongs to CMake/CMakeBuilder, rather than Python. It's CMake knowledge, not Python knowledge :slightly_smiling_face:

**Comment by wdconinc** [comment]
> This sounds like something that has broader applicability than Acts

Totally agree! But this is how most packages seem to deal with this currently, including [`dd4hep`](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/dd4hep/package.py#L201) and [`root`](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/root/package.py#L604). In fact, the reason I stumbled upon this issue is precisely because the included `dd4hep` CMakeConfig in the `acts` build was complaining that the Python version that `acts` had already found disagreed with the Python version used when building `dd4hep`.

I'll also add that this is somewhat package dependent:
- `find_package(Python)` requires `Python_EXECUTABLE`, https://cmake.org/cmake/help/latest/module/FindPython.html
- `find_package(PythonInterp)` requires `PYTHON_EXECUTABLE`, https://cmake.org/cmake/help/latest/module/FindPythonInterp.html (deprecated since CMake 3.12 but still in more use throughout the spack builtin repo)

This means that a central solution needs to handle both cases.

**Comment by adamjstewart** [comment]
I have the same concerns about this as I do for #37967. Not all packages may accept this flag.

**Comment by alalazo** [comment]
> I have the same concerns about this as I do for https://github.com/spack/spack/pull/37967. Not all packages may accept this flag.

Then let's collect the knowledge in a method as proposed in https://github.com/spack/spack/pull/38540#issuecomment-1605299730. If a package needs it, it can call the method, and the details will have a single source of truth.

**Comment by adamjstewart** [comment]
Would invoking that method be any shorter than the changes proposed in this PR?

**Comment by HadrienG2** [comment]
Speaking personally, I think extracting this recurring pattern elsewhere only provides a major benefit to packagers if we can get to the point the point where Python-based CMake packages only need to type `depends_on("python")` to get the correct behavior from CMake.

If this is only a matter of replacing the current hand-written CMake arg in individual packages with a slightly more abstract `self.find_python(self.spec["python"])`, with the caveat mentioned by @wdconinc above that what the abstract method does may depend on obscure details of the caller's CMake configuration, I think the benefit is not that great (still need to remember that Python is special when writing a CMake package), and would then prefer the approach proposed by @wdconinc in this PR for simplicity.

**Comment by alalazo** [comment]
@HadrienG2 Merged this PR while we finish the discussion, since it can be postponed to a following PR. In my opinion:
1. Extracting to a single function in the base class still has the benefit of aggregating knowledge there, even if the packager has to remember calling the function.
2. If there are no cases where using both spelling for the Python executable is problematic, then we can also do it automatically as you proposed.

**Pull request merged**
